node/network/statement-distribution/src/lib.rs

```diff
@@ -771,18 +833,39 @@ async fn circulate_statement_and_dependents(
 
 fn statement_message(relay_parent: Hash, statement: SignedFullStatement)
 	-> protocol_v1::ValidationProtocol
-{
-	protocol_v1::ValidationProtocol::StatementDistribution(
+{ 
+	let msg = if is_statement_large(&statement) {
+		protocol_v1::StatementDistributionMessage::LargeStatement(
+			StatementMetadata {
+				relay_parent,
+				candidate_hash: statement.payload().candidate_hash(),
+				signed_by: statement.validator_index(),
+				signature: statement.signature().clone(),
+			}
+		)
+	} else {
 		protocol_v1::StatementDistributionMessage::Statement(relay_parent, statement)
-	)
+	};
+
+	protocol_v1::ValidationProtocol::StatementDistribution(msg)
+}
+
+/// Check whether a statement should be treated as large statement.
+fn is_statement_large(statement: &SignedFullStatement) -> bool {
+	match &statement.payload() {
+		Statement::Seconded(committed) => 
+			committed.commitments.new_validation_code.is_some(),
```

Comment:
do we want to check the `new_validation_code` size?

Reply:
Yeah, we could do that. Although I would not expect runtime upgrades to be tiny in practice. And even if they are, using request/response would likely still be faster. I am more worried about messages ... those are currently unchecked, hence my question on Element.

Reply:
I will change that check, to calculate real size in total (including all message data).